File: backend/app/parser/merge.py
"""跨块、跨栏、跨页的段落合并。

LaTeX 生成的 PDF 会把一个自然段拆成多个块：实测第 4 页出现
"We evaluate LLMs as prompted three-way classi-" 与 "fiers. The primary..."
两个独立块。若不合并就送去翻译，译文会被拦腰截断且语义丢失。

合并后原块仍保留在页面里（geometry 是渲染层回流所必需的），
只是标记 merged_into 并置 translate=False，由段首块承载完整文本。
"""
from __future__ import annotations

import re

from .extract import stitch_hyphen
from .model import Block, BlockType

# 句末标点：中英文都算。以此结尾说明段落已完结，不应与下一块合并。
_SENT_END = re.compile(r"[.!?;:。！？；：)\]}”’\"']\s*$")
_ENDS_HYPHEN = re.compile(r"[A-Za-z]-$")
# 下一块以小写字母 / 中文 / 闭合括号开头，才可能是续写
_STARTS_CONT = re.compile(r"^[a-z一-鿿),;:\]]")

_MERGEABLE = {BlockType.BODY, BlockType.LIST}


def _can_merge(prev: Block, cur: Block) -> bool:
    # 长标题被排版拆成多块，分别翻译会各自丢掉半句语义，必须先拼回整句
    if prev.type is BlockType.TITLE and cur.type is BlockType.TITLE:
        return True
    if prev.type not in _MERGEABLE or cur.type is not BlockType.BODY:
        return False
    if prev.lang != cur.lang:
        return False
    p, c = prev.text.rstrip(), cur.text.lstrip()
    if not p or not c:
        return False
    # 断词：上块以 "字母-" 结尾，下块小写开头
    if _ENDS_HYPHEN.search(p) and c[:1].islower():
        return True
    # 未完句：上块无句末标点，下块以小写或中文续写
    if not _SENT_END.search(p) and _STARTS_CONT.match(c):
        return True
    return False


def _stitch(head: str, tail: str, hyphen_vocab: set[str], words: set[str]) -> str:
    head, tail = head.rstrip(), tail.lstrip()
    if head.endswith("-") and len(head) >= 2 and head[-2].isalnum():
        return stitch_hyphen(head, tail, hyphen_vocab, words)
    if head and tail and "一" <= head[-1] <= "鿿":
        return head + tail
    return head + " " + tail


def merge_paragraphs(
    ordered: list[Block],
    hyphen_vocab: set[str] | None = None,
    words: set[str] | None = None,
) -> None:
    """就地合并。ordered 必须是全文阅读顺序（跨页连续）。"""
    hyphen_vocab = hyphen_vocab or set()
    words = words or set()
    head: Block | None = None
    for blk in ordered:
        if head is not None and _can_merge(head, blk):
            head.text = _stitch(head.text, blk.text, hyphen_vocab, words)
            head.merged_from.append(blk.id)
            blk.merged_into = head.id
            blk.translate = False
            continue
        head = blk


#: 续行缩进超过首行这么多 pt，即判为同一条文献的续写
_HANGING_INDENT = 4.0


def merge_references(ordered: list[Block], hyphen_vocab=None, words=None) -> None:
    """把悬挂缩进的参考文献条目并成一条。

    文献表用悬挂缩进排版：每条首行齐左，续行缩进。PyMuPDF 会把它们切成
    独立的块，于是一条文献变成 "Helmut Appel, Alexander L Gerlach, and Jan
    Crusius." 与 "2016. The interplay between..." 两块 —— 分开翻译会把
    作者与标题割裂。按每栏内文献块的最左位置识别首行，其余归为续写。
    """
    hyphen_vocab = hyphen_vocab or set()
    words = words or set()

    refs = [b for b in ordered if b.type is BlockType.REFERENCE]
    if not refs:
        return
    # 每（页, 栏）的齐左位置。跨栏、跨页续写的缩进量不同，必须分组求。
    flush: dict[tuple[int, int], float] = {}
    for b in refs:
        key = (b.page, b.column)
        flush[key] = min(flush.get(key, b.bbox[0]), b.bbox[0])

    head: Block | None = None
    for b in ordered:
        if b.type is not BlockType.REFERENCE:
            head = None
            continue
        left = flush.get((b.page, b.column))
        if head is not None and left is not None and b.bbox[0] > left + _HANGING_INDENT:
            head.text = _stitch(head.text, b.text, hyphen_vocab, words)
            head.merged_from.append(b.id)
            b.merged_into = head.id
            b.translate = False
        else:
            head = b

File: web/src/lib/parser/merge.js
/**
 * 跨块、跨栏、跨页的段落合并，以及参考文献条目合并。
 *
 * 一个自然段常被切成多个块（"…three-way classi-" / "fiers. The primary…"），
 * 不合并就送去翻译，译文会被拦腰截断且语义丢失。
 */
import { BlockType } from "./model.js";
import { stitchHyphen } from "./extract.js";

const SENT_END = /[.!?;:。！？；：)\]}”’"']\s*$/;
const ENDS_HYPHEN = /[A-Za-z]-$/;
const STARTS_CONT = /^[a-z一-鿿),;:\]]/;
const MERGEABLE = new Set([BlockType.BODY, BlockType.LIST]);

/** 续行缩进超过首行这么多 pt，即判为同一条文献的续写 */
const HANGING_INDENT = 4;

function canMerge(prev, cur) {
  // 长标题被排版拆成多块，分别翻译会各自丢掉半句语义，必须先拼回整句
  if (prev.type === BlockType.TITLE && cur.type === BlockType.TITLE) return true;
  if (!MERGEABLE.has(prev.type) || cur.type !== BlockType.BODY) return false;
  if (prev.lang !== cur.lang) return false;
  const p = prev.text.replace(/\s+$/, ""), c = cur.text.replace(/^\s+/, "");
  if (!p || !c) return false;
  if (ENDS_HYPHEN.test(p) && /^[a-z]/.test(c)) return true;
  return !SENT_END.test(p) && STARTS_CONT.test(c);
}

function stitch(head, tail, hyphenVocab, words) {
  head = head.replace(/\s+$/, "");
  tail = tail.replace(/^\s+/, "");
  if (head.endsWith("-") && head.length >= 2 && /[A-Za-z0-9]/.test(head[head.length - 2]))
    return stitchHyphen(head, tail, hyphenVocab, words);
  if (head && tail && /[一-鿿]/.test(head.slice(-1))) return head + tail;
  return head + " " + tail;
}

/** 就地合并。ordered 必须是全文阅读顺序（跨页连续）。 */
export function mergeParagraphs(ordered, hyphenVocab, words) {
  let head = null;
  for (const blk of ordered) {
    if (head && canMerge(head, blk)) {
      head.text = stitch(head.text, blk.text, hyphenVocab, words);
      head.mergedFrom.push(blk.id);
      blk.mergedInto = head.id;
      blk.translate = false;
      continue;
    }
    head = blk;
  }
}

/**
 * 把悬挂缩进的参考文献条目并成一条。
 *
 * 文献表用悬挂缩进排版：每条首行齐左，续行缩进。切成独立块后，一条文献变成
 * "Helmut Appel, …Crusius." 与 "2016. The interplay between…" 两块 ——
 * 分开翻译会把作者与标题割裂。按每栏内文献块的最左位置识别首行。
 */
export function mergeReferences(ordered, hyphenVocab, words) {
  const refs = ordered.filter((b) => b.type === BlockType.REFERENCE);
  if (!refs.length) return;

  // 跨栏、跨页续写的缩进量不同，必须按（页, 栏）分组求最左位置
  const flush = new Map();
  for (const b of refs) {
    const key = `${b.page}:${b.column}`;
    flush.set(key, Math.min(flush.get(key) ?? b.bbox[0], b.bbox[0]));
  }

  // 跨栏/跨页续写只可能发生在「上一组的最后一块 → 下一组的第一块」。
  // 不加这个约束，右栏第一条缩进的续行会被并进左栏最后一条文献，
  // 拼出 "…Daphne Ip-Annual Meeting of the Association…" 这种串行。
  const groupOf = (b) => `${b.page}:${b.column}`;
  const lastOf = new Map(), firstOf = new Map();
  for (const b of refs) {
    if (!firstOf.has(groupOf(b))) firstOf.set(groupOf(b), b.id);
    lastOf.set(groupOf(b), b.id);
  }

  let head = null;
  for (const b of ordered) {
    if (b.type !== BlockType.REFERENCE) { head = null; continue; }
    const left = flush.get(groupOf(b));
    const sameGroup = head && groupOf(head) === groupOf(b);
    const flowsOn = head && !sameGroup
      && lastOf.get(groupOf(head)) === head.id && firstOf.get(groupOf(b)) === b.id;
    if (head && (sameGroup || flowsOn) && left !== undefined
        && b.bbox[0] > left + HANGING_INDENT) {
      head.text = stitch(head.text, b.text, hyphenVocab, words);
      head.mergedFrom.push(b.id);
      b.mergedInto = head.id;
      b.translate = false;
    } else {
      head = b;
    }
  }
}
